src/persist-client/src/lib.rs

```diff
@@ -0,0 +1,363 @@
+    /// of a collection. Its most common uses are to recover capabilities that
+    /// have expired (leases) or to attempt to read a TVC that one did not
+    /// create (or otherwise receive capabilities for). If the frontiers have
+    /// been fully released by all other parties, this call may result in
+    /// capabilities with empty frontiers (which are useless).
+    pub async fn open_collection<K, V, T, D>(
+        &self,
+        timeout: Duration,
+        id: Id,
+    ) -> Result<(WriteHandle<K, V, T, D>, ReadHandle<K, V, T, D>), StorageError>
+    where
+        K: Codec,
+        V: Codec,
+        T: Timestamp + Lattice + Codec64,
+        D: Semigroup + Codec64,
+    {
+        let deadline = Instant::now() + timeout;
+
+        let collection = Collection {
+            id,
+            blob: self.blob.clone(),
+            log: Arc::clone(&self.log),
+        };
+        let mut machine = Machine::new(collection);
+
+        let writer_id = WriterId(Uuid::new_v4().as_bytes().to_owned());
+        let reader_id = ReaderId(Uuid::new_v4().as_bytes().to_owned());
+        let (write_cap, read_cap) = machine.register(deadline, &writer_id, &reader_id).await?;
+
+        let write = WriteHandle {
+            machine: machine.clone(),
+            writer_id,
+            _phantom: PhantomData,
+            cap: write_cap,
+        };
+        let read = ReadHandle {
+            machine,
+            reader_id,
+            _phantom: PhantomData,
+            cap: read_cap,
+        };
+
+        Ok((write, read))
+    }
+}
+
+#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct SeqNo(u64);
+
+pub struct CompareAndSet {
+    pub current: SeqNo,
+}
+
+#[async_trait]
+pub trait Log: std::fmt::Debug + Send + 'static {
+    async fn current(&self, deadline: Instant) -> Result<(SeqNo, Option<Vec<u8>>), StorageError>;
+
+    async fn compare_and_set(
+        &self,
+        deadline: Instant,
+        expected: SeqNo,
+        value: Option<Vec<u8>>,
+    ) -> Result<Result<SeqNo, CompareAndSet>, StorageError>;
+
+    async fn compact(&self, deadline: Instant, since: SeqNo) -> Result<(), StorageError>;
+}
+
+#[derive(Debug, Default)]
+pub struct MemLog {
+    entries: Mutex<Vec<(SeqNo, Option<Vec<u8>>)>>,
+}
+
+#[async_trait]
+impl Log for MemLog {
+    async fn current(&self, _deadline: Instant) -> Result<(SeqNo, Option<Vec<u8>>), StorageError> {
+        let current = self
+            .entries
+            .lock()
+            .await
+            .last()
+            .map(|(seqno, entry)| (*seqno, entry.clone()))
+            .unwrap_or_else(|| (SeqNo::default(), None));
+        Ok(current)
+    }
+
+    async fn compare_and_set(
+        &self,
+        deadline: Instant,
+        expected: SeqNo,
+        value: Option<Vec<u8>>,
+    ) -> Result<Result<SeqNo, CompareAndSet>, StorageError> {
+        let (current, _) = self.current(deadline).await?;
```

Comment:
I'm sure you already know this and its totally fine in a prototype but this isn't a correct CAS implementation because two writers can observe `current = x`, and then both write out `x + 1`, which violates the invariant that one of the two writers has to fail. I think its fine to just leave a comment for now?

Reply:
> I think a more general concern I have with this - as someone who is generally ignorant of etcd - how many systems can we implement a CAS on top of? Just want to make sure we're not tying ourselves down to using etcd

a few thoughts
1. no idea. totally flying by the seat of my pants here
2. I'm not even basing this off ectd, I haven't yet gotten to poke at various things and see what they support. keep in mind that this doesn't necessarily imply the system needs to directly have CAS, just that we can performantly build it on top. e.g. it's sorta possible to even do this on top of kafka, where you impose on the log CAS semantics and have to read back your own write to find out if you won
3. see point (1) again :)

Reply:
re: 2 - this is why I asked what systems can we implement CAS on top of :) I think the Kafka thing you're describing will be a bit complicated / not performant in pathological cases because readers/writers can't just read the last value in the topic they have to I think read the entire queue because

imagine the case where `x` is the last item in the queue and also the current seqno.

Now lets say two writers come in, one fast, and one slow, the fast writer completes 100 transactions and writes down (x + 1) observes they won, (x + 2), observes they won ... (x + 100) and observes that they won. Finally, the slow writer writes (x + 1), and observes that they did not win.

so the state of the queue, from oldest to newest insertions looks like:

```
1, 2, 3, ... x, x + 1, x + 2, .... x +100, x +1
```

I think its not sufficient to read any bounded suffix of the queue because there might be arbitrarily many slow writers, and you instead have to read the full queue left to right (oldest to most recent) to determine what the max seqno is.

Reply:
sorry, not being snarky. just not sure how much value there is in discussing the specifics here unless (a) you have a counter-proposal or (b) I've had time to "poke at various things and see what they support"

Reply:
Sure. Concretely: I previously thought you could build a fast CAS implementation on top of an atomic queue like Kafka / Redpanda, and now I don't think that's true. Etcd / CockroachDB both support transactions so this is not an issue for them. That's all I was trying to say.